### PR TITLE
Moved semaphore into a separate header file.

### PR DESCRIPTION
**Pull request by zerodefect**
The semaphore impl is based on Jeff Preshing's semaphore implementation. Nonetheless, it has received some really good changes which has made this synchronisation primitive useful in its own right.

To enable use of this primitive without the having to include blockingconcurrentqueue, the sempahore code has been moved into a separate header file.

Moved semaphore into a separate header file.
Added some preliminary unit tests for LW semaphore.

### CODE REVIEW

**Comment by zerodefect** [comment]
Should I be using tabs or 4 spaces for indentation?  Still a few alignment issues.

**Comment by cameron314** [comment]
Tabs :-)

**Comment by cameron314** [comment]
Although the semaphore might be spaces, come to think of it.

**Comment by zerodefect** [comment]
Ok. Do you mind if I change semaphore code to be consistent?

**Comment by cameron314** [comment]
I guess not. It was done that way so it could be diffed against the original more easily.

**Comment by zerodefect** [comment]
Yeah, that's why I thought I would check.

**Comment by zerodefect** [comment]
Ready for another review.

**Comment by zerodefect** [comment]
Hi @cameron314, I appreciate you're a busy man...any chance for a quick review?

**Comment by cameron314** [comment]
Yup, coming up.

File: lightweightsemaphore.h [hunk 1]
```diff
@@ -0,0 +1,443 @@
+// Provides an efficient implementation of a semaphore (LightweightSemaphore)
+// - a synchronization primitive. This construct has been designed for
```
Comment (cameron314) [inline lightweightsemaphore.h]: I think the "synchronization primitive" part can be left out. It should be fairly familiar to anyone searching for a semaphore...

File: lightweightsemaphore.h [hunk 2]
```diff
@@ -0,0 +1,443 @@
+// Provides an efficient implementation of a semaphore (LightweightSemaphore)
+// - a synchronization primitive. This construct has been designed for 
+// threads and not processes.
+// ©2015-2016 Cameron Desrochers. Distributed under the terms of the simplified
```
Comment (cameron314) [inline lightweightsemaphore.h]: Since this is based on Jeff Preshing's code, I think his code's license should be moved to the top instead. I suppose you could keep a line metnioning that it has been "adapted and extended by Cameron Desrochers", but the important thing is that this is Jeff's code and all the changes I make to it are under the umbrella of his license, not mine.

The LICENSE file will need tweaking too, to reference the semaphore code by the new filename.

File: lightweightsemaphore.h [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+// BSD license, available at the top of concurrentqueue.h.
+
+
```
Comment (cameron314) [inline lightweightsemaphore.h]: Do you like these blank lines or can we remove them?

File: blockingconcurrentqueue.h [hunk 4]
```diff
@@ -7,412 +7,16 @@
 #pragma once
 
 #include "concurrentqueue.h"
+#include "lightweightsemaphore.h"
```
Comment (cameron314) [inline blockingconcurrentqueue.h]: The README will need updating to indicate that the blocking queue now requires the three headers, not just the two.

File: lightweightsemaphore.h [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+#include <cstddef> // For std::size_t
+#include <atomic>
+#include <type_traits> // For std::make_signed<T>
+
+
+
+
+
+// Uses Jeff Preshing's semaphore implementation (under the terms of its
```
Comment (cameron314) [inline lightweightsemaphore.h]: This can be removed once the license is shifted up.

File: lightweightsemaphore.h [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+// freely, subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented; you must not
+//	claim that you wrote the original software. If you use this software
+//	in a product, an acknowledgement in the product documentation would be
+//	appreciated but is not required.
+// 2. Altered source versions must be plainly marked as such, and must not be
+//	misrepresented as being the original software.
+// 3. This notice may not be removed or altered from any source distribution.
+namespace mpmc_sema
```
Comment (cameron314) [inline lightweightsemaphore.h]: No longer required since the lightweight semaphore below is available outside of `mpmc_sema` anyway.

File: lightweightsemaphore.h [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+{
+
+#if defined(_WIN32)
+class Semaphore
+		{
```
Comment (cameron314) [inline lightweightsemaphore.h]: Indentation ;-)

File: lightweightsemaphore.h [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+		// it. So we have to re-adjust the count, but only if the semaphore
+		// wasn't signaled enough times for us too since then. If it was, we
+		// need to release the semaphore too.
+		while (true)
+		{
+			oldCount = m_count.load(std::memory_order_acquire);
+			if (oldCount >= 0 && m_sema.try_wait())
+				return true;
+			if (oldCount < 0 && m_count.compare_exchange_strong(oldCount,
+																oldCount + 1,
```
Comment (cameron314) [inline lightweightsemaphore.h]: Oh boy, I think the autoformatter doesn't like my code :P

File: lightweightsemaphore.h [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+	}
+
+	ssize_t availableApprox() const
+	{
+		ssize_t count = m_count.load(std::memory_order_relaxed);
+		return count > 0 ? count : 0;
+	}
+};
+
+}   // end moodycamel
```
Comment (cameron314) [inline lightweightsemaphore.h]: `end namespace moodycamel`.

Comment (cameron314) [inline lightweightsemaphore.h]: Newline at end of file?

File: tests/unittests/unittests.cpp [hunk 10]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
 		
 		return true;
 	}
+
+	bool acquire_and_signal()
+	{
+		typedef ::moodycamel::LightweightSemaphore semaphore;
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: Capital for type name?

File: tests/unittests/unittests.cpp [hunk 11]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
...
+
+		// Test resource acquisition from one other thread.
+		{
+			semaphore s;
+			s.signal(); // Single resource available
+						
+			const auto fnTestSingleAcquire = 
+				[](::moodycamel::LightweightSemaphore &s) -> bool
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: Use the typedef?

File: tests/unittests/unittests.cpp [hunk 12]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
...
+				{
+					for (std::size_t k = 0; k < 100; ++k)
+					{
+						ASSERT_OR_FAIL(s.wait(100 * 1000)); // Wait 100 ms
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: Hmm, not sure I like this. It could technically fail if the thread is starved for more than 100ms. Can we just wait with an infinite timeout?

Comment (cameron314) [inline tests/unittests/unittests.cpp]: `ASSERT_OR_FAIL` is really just a fancy macro that does `return false` on failure. (My test framework is pretty basic.) This won't work from within a thread callback, only the test body itself. Instead, an atomic boolean should be set and assertion on after the threads join. (The return value from the `std::thread` entry function cannot be recuperated.)

File: tests/unittests/unittests.cpp [hunk 13]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
 		
 		return true;
 	}
+
+	bool acquire_and_signal()
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: It would be amazing if we could get the semaphore tests to run under Relacy. Maybe we could simulate the underlying OS semaphore with a C++11 concurrency primitive that Relacy understands?

File: tests/unittests/unittests.cpp [hunk 14]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
...
+			t1.join();
+			t2.join();
+			
+			ASSERT_OR_FAIL(1==s.availableApprox());
+		}
+
+		// Test resource acquisition from multiple threads.
+		{
+			const std::size_t ITERATIONS = 100;
+			const std::array<std::size_t, 2> arrayItemsToWait = {2,3};
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: Why not just a plain array? Does it really need have `array` in its name?

File: tests/unittests/unittests.cpp [hunk 15]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
...
+			semaphore s;
+
+			s.signal((100 * arrayItemsToWait[0]) + (100 * arrayItemsToWait[1]));
+			
+			const auto fnTestMultiAcquire =
+				[ITERATIONS](const std::size_t  n, ::moodycamel::LightweightSemaphore &s) -> bool
+				{
+					for (std::size_t k = 0; k < ITERATIONS; ++k)
+					{
+						const ssize_t signedN = static_cast<ssize_t>(n);
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: Is `ssize_t` in scope here? How?

File: tests/unittests/unittests.cpp [hunk 16]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
...
+			std::future<bool> f1 = task1.get_future();
+			std::future<bool> f2 = task2.get_future();
+
+			std::thread t1(std::move(task1), arrayItemsToWait[0], std::ref(s));
+			std::thread t2(std::move(task2), arrayItemsToWait[1], std::ref(s));
+
+			t1.join();
+			t2.join();
+			
+			ASSERT_OR_FAIL(f1.get());
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: Since the assertion message was already printed out at this point, this should be a simply `if (!...) return false;` instead.

File: tests/unittests/unittests.cpp [hunk 17]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
...
+
+		// Test resource acquisition from one other thread.
+		{
+			semaphore s;
+			s.signal(); // Single resource available
+						
+			const auto fnTestSingleAcquire = 
+				[](::moodycamel::LightweightSemaphore &s) -> bool
+				{
+					for (std::size_t k = 0; k < 100; ++k)
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: 100 iterations is likely not enough to force concurrency (the overhead of starting a new thread will eclipse the time the thread takes to execute).

File: tests/unittests/unittests.cpp [hunk 18]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
...
+			// Ideally std::thread would have 'try_join_for' like Boost :(
+			t1.join();
+			t2.join();
+			
+			ASSERT_OR_FAIL(1==s.availableApprox());
+		}
+
+		// Test resource acquisition from multiple threads.
+		{
+			const std::size_t ITERATIONS = 100;
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: Need moar :-)

File: tests/unittests/unittests.cpp [hunk 19]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
...
+				};
+				
+			std::packaged_task<bool(std::size_t, moodycamel::LightweightSemaphore&)> task1(fnTestMultiAcquire);
+			std::packaged_task<bool(std::size_t, moodycamel::LightweightSemaphore&)> task2(fnTestMultiAcquire);
+			
+			std::future<bool> f1 = task1.get_future();
+			std::future<bool> f2 = task2.get_future();
+
+			std::thread t1(std::move(task1), arrayItemsToWait[0], std::ref(s));
+			std::thread t2(std::move(task2), arrayItemsToWait[1], std::ref(s));
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: Ideally there would be more than just two threads hammering on the semaphore. Also, this test is fine in and of itself, but the case where the semaphore dips below zero and bounces back is where all the tricky logic lies, and is not tested at all.

File: tests/unittests/unittests.cpp [hunk 20]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
...
+			t1.join();
+			t2.join();
+			
+			ASSERT_OR_FAIL(f1.get());
+			ASSERT_OR_FAIL(f2.get());
+			
+			ASSERT_OR_FAIL(0==s.availableApprox());
+		}
+
+		semaphore s;
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: Maybe this simple test should go to the top of the function, before the heavy duty threaded tests execute?

File: tests/unittests/unittests.cpp [hunk 21]
```diff
@@ -4264,6 +4269,137 @@ class ConcurrentQueueTests : public TestClass<ConcurrentQueueTests>
...
+		ASSERT_OR_FAIL(s.availableApprox()==10);
+		s.waitMany(10);
+		ASSERT_OR_FAIL(s.availableApprox()==0);
+
+		return true;
+	}
+
+	bool try_acquire_and_signal()
+	{
+		typedef ::moodycamel::LightweightSemaphore semaphore;
```
Comment (cameron314) [inline tests/unittests/unittests.cpp]: Should test with threads too.

File: blockingconcurrentqueue.h (hunk 4 above)
Comment (zerodefect) [reply blockingconcurrentqueue.h]: Done.

File: lightweightsemaphore.h (hunk 6 above)
Comment (zerodefect) [reply lightweightsemaphore.h]: This confused me a bit. Are you suggesting that I remove the zlib docs? Can you see what I currently have :) ?

File: lightweightsemaphore.h (hunk 8 above)
Comment (zerodefect) [reply lightweightsemaphore.h]: Done.

**Comment by zerodefect** [comment]
Not fixed the unit tests yet.  I want to get lightweightsemaphore.h correct before I move forward. I had a few questions about your license formatting requests.

**Comment by Kronuz** [comment]
Any news about it? Will this be merged soon?
